core/domain/question_services.py

```diff
@@ -201,3 +203,39 @@ def update_question(committer_id, question_id, change_list, commit_message):
     """
     question = apply_change_list(question_id, change_list)
     _save_question(committer_id, question, change_list, commit_message)
+
+
+def get_questions_batch(
+        collection_id, skill_ids, user_id, question_play_counts, batch_size):
+    """Fetches a batch of questions.
+
+    Args:
+        collection_id: str. Id of the collection.
+        skill_ids: list(str). A list of skill ids.
+        user_id: str. Id of the user.
+        question_play_counts: dict. A dict mapping question ids to the number
+            of times it has been played in this question session.
+        batch_size: int. The intende number of questions to be returned.
+
+    Returns:
+        A list of Question objects.
+    """
+    user_skill_ids = collection_services.get_acquired_skills_of_user(user_id)
+    question_skill_ids = list(set(user_skill_ids) & set(skill_ids))
+
+    collection = collection_services.get_collection_by_id(collection_id)
+    question_ids = []
+    for skill_id in question_skill_ids:
+        if collection.skills[skill_id]:
```

Comment:
Why is this check needed?

Reply:
There maybe skills which are common to skill list provided and user skills but the common skill may not be present in collection. I am guessing your response would be why would you fetch questions which are not related to Collection.
Since skill list is provided to this method. I am not trusting it to be correct. Correct me if I am wrong.

Reply:
Hm, I think having the check is fine. But I guess you should do `if skill_id in collection.skills`, then? Otherwise you'll get a key error.

Maybe add a test for that case, too.

Reply:
Updated, test pending.